File: tlg/src/main/java/ch/bailu/tlg/MatrixWithShape.java
package ch.bailu.tlg;

import java.io.BufferedInputStream;
import java.io.BufferedOutputStream;
import java.io.IOException;

public class MatrixWithShape extends MatrixManipulator {

    private MatrixShape movingShape;

    public MatrixWithShape(int w, int h) {
        super(w, h);
        movingShape = new MatrixShape();
    }

    public MatrixWithShape(BufferedInputStream reader) throws IOException {
        super(reader);
        movingShape = new MatrixShape(reader);
    }

    public void writeState(BufferedOutputStream output) throws IOException {
        super.writeState(output);
        movingShape.writeState(output);
    }



    public boolean setMovingShape(MatrixShape shape) {
        movingShape = shape;
        return autoPlaceShape(movingShape);
    }

    public MatrixShape getMovingShape() {
        return movingShape;
    }

    public void removeMovingShape() {
        removeShape(movingShape);
    }

    public void setRandomMovingShape(int level, int color, int turn) {
        MatrixShape shape = new MatrixShape();
        shape.initializeFormAndColor(level, color);
        movingShape = shape.getRotatedCopy(turn);
        autoPlaceShape(movingShape);
    }


    public void moveShapeLeft() {
        TlgPoint pos = movingShape.getPos();
        pos.x -= 1;
        moveShape(movingShape,pos);
    }


    public void moveShapeRight() {
        TlgPoint pos = movingShape.getPos();
        pos.x += 1;
        moveShape(movingShape,pos);

    }

    public boolean moveShapeDown() {
        TlgPoint pos = movingShape.getPos();
        pos.y += 1;
        return moveShape(movingShape,pos);
    }

    public void moveShapeTurn() {
       MatrixShape temp = movingShape.getRotatedCopy(1);

       removeShape(movingShape);
       if (placeShape(temp)) movingShape=temp;
       else placeShape(movingShape);
    }


    public void centerMovingShape() {
        TlgRectangle r = movingShape.getActiveArea();
        int x = ( (getWidth() - r.getWidth()) / 2 ) - r.left;
        int y = ( (getHeight() - r.getHeight()) / 2 ) - r.top;

        if (x != 0 || y != 0) {
            moveShape(movingShape,x, y);
        }
    }
}

File: tlg/src/main/kotlin/matrix/MatrixWithShape.kt
package matrix

import geometry.TlgPoint
import geometry.TlgRectangle
import java.io.BufferedInputStream
import java.io.BufferedOutputStream
import java.io.IOException

open class MatrixWithShape : MatrixManipulator {
    var movingShape: MatrixShape
        private set

    constructor(w: Int, h: Int) : super(w, h) {
        movingShape = MatrixShape()
    }

    constructor(reader: BufferedInputStream) : super(reader) {
        movingShape = MatrixShape(reader)
    }

    @Throws(IOException::class)
    override fun writeState(out: BufferedOutputStream) {
        super.writeState(out)
        movingShape.writeState(out)
    }

    fun setMovingShape(shape: MatrixShape): Boolean {
        movingShape = shape
        return autoPlaceShape(movingShape)
    }

    fun removeMovingShape() {
        removeShape(movingShape)
    }

    fun setRandomMovingShape(level: Int, color: Int, turn: Int) {
        val shape = MatrixShape()
        shape.initializeFormAndColor(level, color)
        movingShape = shape.getRotatedCopy(turn)
        autoPlaceShape(movingShape)
    }

    fun moveShapeLeft() {
        val pos: TlgPoint = movingShape.pos
        pos.x -= 1
        moveShape(movingShape, pos)
    }

    fun moveShapeRight() {
        val pos: TlgPoint = movingShape.pos
        pos.x += 1
        moveShape(movingShape, pos)
    }

    fun moveShapeDown(): Boolean {
        val pos: TlgPoint = movingShape.pos
        pos.y += 1
        return moveShape(movingShape, pos)
    }

    fun moveShapeTurn() {
        val temp = movingShape.getRotatedCopy(1)
        removeShape(movingShape)
        if (placeShape(temp)) movingShape = temp else placeShape(movingShape)
    }

    fun centerMovingShape() {
        val r: TlgRectangle = movingShape.activeArea
        val x: Int = (width - r.width) / 2 - r.left
        val y: Int = (height - r.height) / 2 - r.top
        if (x != 0 || y != 0) {
            moveShape(movingShape, x, y)
        }
    }
}
